riscv/src/register/macros.rs

```diff
@@ -600,3 +600,445 @@ macro_rules! clear_pmp {
+     $(#[$set_field_doc:meta])+
+     $set_field:ident,
+     $field_ty:ident,
+     range: [$field_start:literal : $field_end:literal],
+    ) => {
+         $crate::read_only_csr_field!(
+             $ty,
+             $(#[$field_doc])+
+             $field,
+             $(#[$try_field_doc])+
+             $try_field,
+             $field_ty,
+             range: [$field_start : $field_end],
+         );
+
+         $crate::write_only_csr_field!(
+             $ty,
+             $(#[$set_field_doc])+
+             $set_field,
+             $field_ty,
+             range: [$field_start : $field_end],
+         );
+    };
+}
+
+/// Defines field accesor functions for a read-only CSR type.
+#[macro_export]
+macro_rules! read_only_csr_field {
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident: $bit:literal$(,)?) => {
+        const _: () = assert!($bit < usize::BITS);
+
+        impl $ty {
+            $(#[$field_doc])+
+            #[inline]
+            pub fn $field(&self) -> bool {
+                $crate::bits::bf_extract(self.bits, $bit, 1) != 0
+            }
+        }
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$try_field_doc:meta])+
+     $try_field:ident,
+     range: $bit_start:literal..=$bit_end:literal$(,)?) => {
+        const _: () = assert!($bit_end < usize::BITS);
+        const _: () = assert!($bit_start < $bit_end);
+
+        impl $ty {
+            $(#[$field_doc])+
+            #[inline]
+            pub fn $field(&self, index: usize) -> bool {
+                self.$try_field(index).unwrap()
+            }
+
+            $(#[$try_field_doc])+
+            #[inline]
+            pub fn $try_field(&self, index: usize) -> $crate::result::Result<bool> {
+                if ($bit_start..=$bit_end).contains(&index) {
+                    Ok($crate::bits::bf_extract(self.bits, index, 1) != 0)
+                } else {
+                    Err($crate::result::Error::IndexOutOfBounds {
+                        index,
+                        min: $bit_start,
+                        max: $bit_end,
+                    })
+                }
+            }
+        }
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident: [$bit_start:literal : $bit_end:literal]$(,)?) => {
+        const _: () = assert!($bit_end < usize::BITS);
+        const _: () = assert!($bit_start < $bit_end);
+
+        impl $ty {
+            $(#[$field_doc])+
+            #[inline]
+            pub fn $field(&self) -> usize {
+                $crate::bits::bf_extract(self.bits, $bit_start, $bit_end - $bit_start + 1)
+            }
+        }
+    };
+
+    ($ty:ident,
+     $(#[$field_doc:meta])+
+     $field:ident,
+     $(#[$try_field_doc:meta])+
+     $try_field:ident,
+     $field_ty:ident,
+     range: [$field_start:literal : $field_end:literal]$(,)?
+    ) => {
+        const _: () = assert!($field_end < usize::BITS);
+        const _: () = assert!($field_start <= $field_end);
```

Comment:
For the enum branches, I left the `<=` check.

My thinking here is that there may be CSR definitions where users want to use an enum to clarify the meaning of a single-bit field.

For example, a platform with a custom CSR where a single-bit field can mean:

- `0`: set tolerance to custom value X
- `1`: set tolerance to custom value Y

And the enum definition would be something like:

```rust
csr_field_enum! {
   /// Some custom CSR field.
   CsrFieldTolerance,
   range: [0:0],
   default: ToleranceX,
   ToleranceX = 0,
   ToleranceY = 1,
}
```